src/main/java/org/opensearch/geospatial/ip2geo/cache/Ip2GeoCache.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.geospatial.ip2geo.cache;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import lombok.Getter;
+import lombok.extern.log4j.Log4j2;
+
+import org.opensearch.common.xcontent.LoggingDeprecationHandler;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.geospatial.ip2geo.common.DatasourceFacade;
+import org.opensearch.geospatial.ip2geo.common.DatasourceState;
+import org.opensearch.geospatial.ip2geo.jobscheduler.Datasource;
+import org.opensearch.index.engine.Engine;
+import org.opensearch.index.shard.IndexingOperationListener;
+import org.opensearch.index.shard.ShardId;
+
+@Log4j2
+public class Ip2GeoCache implements IndexingOperationListener {
+    private final DatasourceFacade datasourceFacade;
+    private Map<String, DatasourceMetadata> data;
+
+    public Ip2GeoCache(final DatasourceFacade datasourceFacade) {
+        this.datasourceFacade = datasourceFacade;
+    }
+
+    public String getIndexName(final String datasourceName) {
+        return getData().getOrDefault(datasourceName, DatasourceMetadata.EMPTY_METADATA).getIndexName();
+    }
+
+    public boolean isExpired(final String datasourceName) {
+        return getData().getOrDefault(datasourceName, DatasourceMetadata.EMPTY_METADATA).getExpirationDate().isBefore(Instant.now());
```

Comment:
why not to fail fast instead of returning some data based on empty_metadata? If we pretend that datasource exists then it may lead to errors on the client side if we return something from "isExpired" or "getIndexName"

Reply:
I choose not to throw exception but return default value. There is no risk of pretending datasource exist by this method because this method should not be used to check if datasource exist or not. There is another method to check that. This just simply the code with low risk. If you still think throwing exception is better, I can make a change.